### ISSUE TITLE
Mitigate the possibility of detecting the incognito mode through storage.estimate()

### ISSUE THREAD

**Issue by uazo** [bromite/bromite#730]
### Is your feature request related to privacy?

Yes

### Is there a patch available for this feature somewhere?

I have not found anything

### Describe the solution you would like

due to a bug in navigator.storage.estimate() it is possible to know if a session is in incognito mode.

some sites like they use
```
x = navigator.storage.estimate().quota;
!! (x && x <12e7)
```
to understand if the session is anonymous.
way to reproduce it https://github.com/bromite/bromite/issues/695#issuecomment-677221366
same bug can be found in the windows version.

issue is present in crbug since May 6, 2019
https://bugs.chromium.org/p/chromium/issues/detail?id=959839
https://bugs.chromium.org/p/chromium/issues/detail?id=990731

### Describe alternatives you have considered

1) change https://source.chromium.org/chromium/chromium/src/+/master:storage/browser/quota/quota_settings.cc
2) enable IncognitoDynamicQuota flag (default in upstream)
3) consider changing of IncognitoQuotaRatioLowerBound e IncognitoQuotaRatioUpperBound, to make them as truthful as possible

```
diff --git a/storage/browser/quota/quota_features.cc b/storage/browser/quota/quota_features.cc
index 8e47aef9c4..3ce78ec278 100644
--- a/storage/browser/quota/quota_features.cc
+++ b/storage/browser/quota/quota_features.cc
@@ -27,7 +27,7 @@ const base::Feature kQuotaUnlimitedPoolSize{"QuotaUnlimitedPoolSize",
 // IncognitoDynamicQuota enables dynamic assignment of quota to incognito mode
 // based on the physical memory size and removes the fixed upper cap for it.
 const base::Feature kIncognitoDynamicQuota{"IncognitoDynamicQuota",
-                                           base::FEATURE_DISABLED_BY_DEFAULT};
+                                           base::FEATURE_ENABLED_BY_DEFAULT};

 // Dynamic quota for incognito mode would be set by a random fraction of
 // physical memory, between |IncognitoQuotaRatioLowerBound| and
@@ -35,7 +35,7 @@ const base::Feature kIncognitoDynamicQuota{"IncognitoDynamicQuota",
 constexpr base::FeatureParam<double> kIncognitoQuotaRatioLowerBound{
     &kIncognitoDynamicQuota, "IncognitoQuotaRatioLowerBound", 0.1};
 constexpr base::FeatureParam<double> kIncognitoQuotaRatioUpperBound{
-    &kIncognitoDynamicQuota, "IncognitoQuotaRatioUpperBound", 0.2};
+    &kIncognitoDynamicQuota, "IncognitoQuotaRatioUpperBound", 0.9};

 }  // namespace features
 }  // namespace storage
diff --git a/storage/browser/quota/quota_settings.cc b/storage/browser/quota/quota_settings.cc
index d5f6709552..7b57a4c13d 100644
--- a/storage/browser/quota/quota_settings.cc
+++ b/storage/browser/quota/quota_settings.cc
@@ -38,7 +38,7 @@ QuotaSettings CalculateIncognitoDynamicSettings(
   // The incognito pool size is a fraction of the amount of system memory,
   // and the amount is capped to a hard limit.
   double incognito_pool_size_ratio = 0.1;  // 10%
-  int64_t max_incognito_pool_size = 300 * kMBytes;
+  int64_t max_incognito_pool_size = 30000 * kMBytes;
   if (base::FeatureList::IsEnabled(features::kIncognitoDynamicQuota)) {
     const double lower_bound = features::kIncognitoQuotaRatioLowerBound.Get();
     const double upper_bound = features::kIncognitoQuotaRatioUpperBound.Get();
@@ -67,10 +67,10 @@ base::Optional<QuotaSettings> CalculateNominalDynamicSettings(
   base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
                                                 base::BlockingType::MAY_BLOCK);

-  if (is_incognito) {
-    return CalculateIncognitoDynamicSettings(
-        device_info_helper->AmountOfPhysicalMemory());
-  }
+//  if (is_incognito) {
+//    return CalculateIncognitoDynamicSettings(
+//        device_info_helper->AmountOfPhysicalMemory());
+//  }

   // The fraction of the device's storage the browser is willing to use for
   // temporary storage.
@@ -124,6 +124,10 @@ base::Optional<QuotaSettings> CalculateNominalDynamicSettings(
   QuotaSettings settings;

   int64_t total = device_info_helper->AmountOfTotalDiskSpace(partition_path);
+  if (is_incognito) {
+    auto randomvalue = CalculateIncognitoDynamicSettings(total);
+    total = randomvalue.pool_size;
+  }
   if (total == -1) {
     LOG(ERROR) << "Unable to compute QuotaSettings.";
     return base::nullopt;
```

i tried and it seems to work.
although in incognito mode the associated db is kept in memory, I have not found any problems related to memory allocations.

Another alternative is to just change the result in blink here
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/quota/deprecated_storage_quota.cc;l=90
making it always greater than that value.

**Comment by csagan5** [comment]
> although in incognito mode the associated db is kept in memory, I have not found any problems related to memory allocations.

You need to test with a page that actually allocates the space to verify this.

> due to a bug in navigator.storage.estimate() it is possible to know if a session is in incognito mode.
> 
> some sites like they use

I don't think it's a bug, it is designed to work like that. Do we have a list of sites to use to test the problem? Maybe the upstream Chromium issues mention some? It would be best that we can produce such list to use it to run the tests.

You mention multiple approaches here, and a combination of them. We need to check each of them.

The best would be that upstream fixes handles this problem; until that, let's start from step 1: list of sites to use to test the problem and possibly what is the technique (we need to extract Javascript snippets).

**Comment by uazo** [comment]
>  1: list of sites to use to test the problem and possibly what is the technique (we need to extract Javascript snippets).

sure we can post publicly?
I already have a script, but of course I had to "decompile" it, could it be a problem?

**Comment by csagan5** [comment]
It's not a problem, but are you sure these are not already publicly known techniques? I would think they have minified existing open source libraries.

**Comment by csagan5** [comment]
@uazo first thing I would like to test is this patch alone:
```diff
diff --git a/storage/browser/quota/quota_features.cc b/storage/browser/quota/quota_features.cc
index 8e47aef9c494..96e8ffe20cba 100644
--- a/storage/browser/quota/quota_features.cc
+++ b/storage/browser/quota/quota_features.cc
@@ -27,15 +27,15 @@ const base::Feature kQuotaUnlimitedPoolSize{"QuotaUnlimitedPoolSize",
 // IncognitoDynamicQuota enables dynamic assignment of quota to incognito mode
 // based on the physical memory size and removes the fixed upper cap for it.
 const base::Feature kIncognitoDynamicQuota{"IncognitoDynamicQuota",
-                                           base::FEATURE_DISABLED_BY_DEFAULT};
+                                           base::FEATURE_ENABLED_BY_DEFAULT};
 
 // Dynamic quota for incognito mode would be set by a random fraction of
 // physical memory, between |IncognitoQuotaRatioLowerBound| and
 // |IncognitoQuotaRatioUpperBound|.
 constexpr base::FeatureParam<double> kIncognitoQuotaRatioLowerBound{
-    &kIncognitoDynamicQuota, "IncognitoQuotaRatioLowerBound", 0.1};
+    &kIncognitoDynamicQuota, "IncognitoQuotaRatioLowerBound", 0.15};
 constexpr base::FeatureParam<double> kIncognitoQuotaRatioUpperBound{
-    &kIncognitoDynamicQuota, "IncognitoQuotaRatioUpperBound", 0.2};
+    &kIncognitoDynamicQuota, "IncognitoQuotaRatioUpperBound", 0.3};
 
 }  // namespace features
 }  // namespace storage
```

Q: does this change still allow to detect incognito mode?

Note that upstream is working on something new called "storage pressure", see https://source.chromium.org/chromium/chromium/src/+/master:storage/browser/quota/quota_features.cc;bpv=1;bpt=1;l=41?q=IncognitoDynamicQuota&ss=chromium%2Fchromium%2Fsrc and https://bugs.chromium.org/p/chromium/issues/detail?id=1096549

**Comment by uazo** [comment]
> but are you sure these are not already publicly known techniques? 

the one I have seen is based on this bug.

>  I would think they have minified existing open source libraries.

I think it is closed source instead, at least so it seems from some things I read in the decompiled source.
including even a hack for (internal) programmers to bypass that detection!

> Q: does this change still allow to detect incognito mode?

I think so.
I think the real problem is here
```
    return CalculateIncognitoDynamicSettings(
        device_info_helper->AmountOfPhysicalMemory());
```
as you can see, in incognito mode the value is the amount of RAM, not the disk space

> Note that upstream is working on something new called "storage pressure"

this i think the implementation
https://source.chromium.org/chromium/chromium/src/+/master:storage/browser/quota/quota_manager.cc;l=1428
it seems to be related only to disk space, if too little, it starts deleting by origin.

> Note that upstream is working on something new called "storage pressure"

that's why I checked if there can be ways of figuring out incognito mode with my patch too, asking too much memory.

I thought that if you try to allocate too much memory, incognito mode might fail. If that happens, we can edit here
https://source.chromium.org/chromium/chromium/src/+/master:storage/browser/quota/quota_database.cc;l=571
We open the db really on disk also for the incognito mode, and we insert a flag in the table that indicates that mode (as I saw in the cookie store when I checked the session_only) and we delete the records at each reopening.

or maybe it's really better to just act in the render process.

**Comment by csagan5** [comment]
> > but are you sure these are not already publicly known techniques?
> 
> the one I have seen is based on this bug.

The article referenced in the Chromium issue is http://mishravikas.com/articles/2019-07/bypassing-anti-incognito-detection-google-chrome.html and it mentions the technique, so it's publicly known, nothing secret.

> I think it is closed source instead, at least so it seems from some things I read in the decompiled source.
including even a hack for (internal) programmers to bypass that detection!

Interesting, but it's not a secret technique. See above.

> as you can see, in incognito mode the value is the amount of RAM, not the disk space

Yes, I am aware. It is by design.

> this i think the implementation
> https://source.chromium.org/chromium/chromium/src/+/master:storage/browser/quota/quota_manager.cc;l=1428
> it seems to be related only to disk space, if too little, it starts deleting by origin.

If you read both issues you linked, the pressure problem is mentioned to detect abuse of the feature (in both incognito and normal mode, when incognito would start using encrypted disk as storage). That seems to be where upstream is headed with this.


> that's why I checked if there can be ways of figuring out incognito mode with my patch too, asking too much memory.

I prefer to wait on upstream for a proper solution; until that we can enable the patch I pasted in previous comment, or maybe use in `CalculateIncognitoDynamicSettings` the free disk storage; if they try to write it will of course fail, but I am fine with that mode of detection until upstream finishes the work on encrypted storage etc.

**Comment by uazo** [comment]
> I prefer to wait on upstream for a proper solution

certainly

> until that we can enable the patch I pasted in previous comment

ok, I tried it, it seems to work, but it may not be enough

> maybe use in CalculateIncognitoDynamicSettings the free disk storage

as you prefer

> if they try to write it will of course fail,

Not necessarily. bydesign access can be denied by the user or by the content settings, the website may still work

**Comment by csagan5** [comment]
> ok, I tried it, it seems to work, but it may not be enough

The future improvements will be covered by upstream e.g. when publishers start to change their scripts.

**Comment by csagan5** [comment]
Fixed in `85.0.4183.110`.
